[Caffe2] Enhance test for CollectAndDistributeOp

pooyadavoodi opened the pull request:
This also changes the operator and the test to use stable sort
otherwise the test will fail due to differences between the op
and the test when facing ROIs of the same score.

pooyadavoodi committed:
Caffe2: Enhance test for CollectAndDistributeOp

This also changes the operator and the test to use stable sort
otherwise the test will fail due to differences between the op
and the test when facing ROIs of the same score.

yinghai commented on caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc:
```diff
@@ -58,12 +58,11 @@ void SortAndLimitRoIsByScores(Eigen::Ref<const EArrXf> scores, int n,
   // Note that people have found nth_element + sort to be much faster
   // than partial_sort so we use it here
   if (n > 0 && n < rois.rows()) {
-    std::nth_element(idxs.begin(), idxs.begin() + n, idxs.end(), comp);
```
Why did we remove this?

orionr replied:
I agree - we should keep this. The comment I made above about nth_element + sort hasn't really been tested on real-world cases, so it could likely be wrong. If you could verify given the larger range of test cases below, that would be great. Then we can decide whether to use nth_element + sort or partial_sort.

In both cases, though, we can make it stable by adjusting comp to also test index:

```
auto comp = [&scores](int lhs, int rhs) {
  if (scores(lhs) > scores(rhs)) return true;
  if (scores(lhs) < scores(rhs)) return false;
  return lhs > rhs;
};
```
Then there's no need to use stable_sort. Thoughts?

yinghai commented:
And also, why is stable sort a requirement other than helping with the tests?

orionr commented on caffe2/python/operator_test/collect_and_distribute_fpn_rpn_proposals_op_test.py:
```diff
@@ -118,72 +120,93 @@ def collect_and_distribute_fpn_rpn_ref(*inputs):
 
 class TestCollectAndDistributeFpnRpnProposals(hu.HypothesisTestCase):
-    def run_on_device(self, device_opts):
+    @given(proposal_count=st.integers(min_value=1000, max_value=8000),
```
Great test coverage range - thank you for this!

orionr reviewed (changes requested):
This is great @pooyadavoodi! Let's discuss the stable sort before this comes in. It would be great, given your new test coverage, to see whether nth_element + sort is indeed better than partial_sort.

pooyadavoodi committed:
Caffe2: Adjust comparator to make std::nth_element and std::sort stable

Revert the removal of std::nth_element and std::sort and adding of
std::stable_sort.

pooyadavoodi commented:
I reverted removing of `std::nth_element` and `std::sort`. Changing the comparator solves the problem.

The only reason I had removed `std::nth_element` and `std::sort` and used `std::stable_sort` was because of the test, and that problem is solved now after changing the comparator.

yinghai reviewed (approved):
Cool. I think `nth_element` and `partial_sort` comparison is out of the scope of this PR? Although I think `partial_sort` is more efficient, if we don't really have a use case, let's keep it this way.

pooyadavoodi commented:
I agree, we can compare them later. My focus currently is on writing a CUDA implementation for this operator, and that's why I wanted to improve the test first for the sake of my new implementation.

orionr reviewed (approved):
Agreed - let's defer `nth_element` vs `partial_sort` for now. @yinghai are the failing tests as expected or should we wait to land?

orionr commented:
Looks like these tests are failing on other PRs - merging.

yinghai commented:
Yeah, the two failures are not related to this PR.
